File: JAVA/CountEqualAndDivisiblePairsInAnArray.java
/**
 * @version JAVA
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package JAVA;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CountEqualAndDivisiblePairsInAnArray {

    public static void main(String[] args) {
        System.out.printf("Resultado: %d\n", countPairs( new int[]{1,2}, 3 ));
    }

    public static int countPairs(int[] nums, int k) {
        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            map.computeIfAbsent(nums[i], v -> new ArrayList<>()).add(i);
        }

        int result = 0;
        for (List<Integer> indices : map.values()) {
            for (int i = 0; i < indices.size() - 1; i++) {
                for (int j = i + 1; j < indices.size(); j++) {
                    if ((indices.get(i) * indices.get(j)) % k == 0) {
                        result++;
                    }
                }
            }
        }

        return result;
    }
}

File: GO/CountEqualAndDivisiblePairsInAnArray.go
/**
 * @version GO 1.23.5
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

func countPairs(nums []int, k int) int {
	m := make(map[int][]int)
	for i, v := range nums {
		m[v] = append(m[v], i)
	}

	r := 0
	for _, is := range m {
		for i := 0; i < len(is)-1; i++ {
			for j := i + 1; j < len(is); j++ {
				if is[i]*is[j]%k == 0 {
					r++
				}
			}
		}
	}
	return r
}
